### PR TITLE
ENH: special: add `softplus`

### LINKED ISSUE #17905: ENH: Add softplus implementation
### Is your feature request related to a problem? Please describe.

It might be nice to have a numerically stable [softplus](https://en.wikipedia.org/wiki/Rectifier_(neural_networks)#Softplus) implementation, ie `np.log1p(np.exp(x))`

### Describe the solution you'd like.

Something based on one of these stackoverflow answers:

- https://cs.stackexchange.com/questions/110798/numerically-stable-log1pexp-calculation
- https://stackoverflow.com/questions/44230635/avoid-overflow-with-softplus-function-in-python

### Describe alternatives you've considered.

I could use `logsumexp` with zeros in the first argument
```
>>> from scipy.special import logsumexp
>>> import numpy as np
>>> logsumexp(np.vstack((np.zeros(10), 10**np.arange(10))), axis=0)
```
but it's a bit cumbersome. Not sure if it's enough of a pain to warrant this enhancement proposal though.

### Additional context (e.g. screenshots, GIFs)

_No response_

### PR DESCRIPTION
**Pull request by aadya940**
#### Reference issue
Closes #17905 

#### What does this implement/fix?
This Pull Request implements a numerically stable version of the softplus function by wrapping the logaddexp function of the numpy API. All test cases have been covered.

#### Additional information
No.

### CODE REVIEW

File: scipy/special/_basic.py
Comment (tupui) [inline scipy/special/_basic.py:3183]:
```suggestion
    ----------
    x : array_like
        Input array-like object.

    **kwargs :
        Additional keyword arguments to be passed to `np.logaddexp`.

    Returns
    -------
    softplus : array_like
        Array containing the softplus of each element in `x`.

    Examples
    --------
    >>> softplus(0)
    0.6931471805599453

    >>> softplus([-1, 0, 1])
    array([0.31326169, 0.69314718, 1.31326169])
```

File: scipy/special/_basic.py
Comment (tupui) [inline scipy/special/_basic.py:3161]:
```suggestion
    The softplus function is defined as: ``softplus(x) = log(1 + exp(x))``.
```

File: scipy/special/tests/test_softplus.py
Comment (tupui) [inline scipy/special/tests/test_softplus.py:1]: I am not sure about creating a new file just for this function. I would put this in an existing test file.

File: scipy/special/tests/test_softplus.py
Comment (tupui) [inline scipy/special/tests/test_softplus.py:11]: The test could be parametrised.

Where are the values coming from? It would be good if we had a reference to cite or something else.

**Review by tupui — DISMISSED** [review]
Thank you @aadya940 for the PR. I have a few suggestions. For the rest I will let @WarrenWeckesser and others take over

File: scipy/special/_basic.py
Comment (tupui) [inline scipy/special/_basic.py:78]: The list is alphabetically ordered, should be moved above.

**Comment by lorentzenchr** [comment]
Is the naming already decided? While pytorch calls it `softplus`, it‘s not a common name that I have met often. Something like `log1pexp` is also a bit cryptic. I just wanted to make sure, it‘s a deliberate decision.

**Comment by tupui** [comment]
> Is the naming already decided? While pytorch calls it `softplus`, it‘s not a common name that I have met often. Something like `log1pexp` is also a bit cryptic. I just wanted to make sure, it‘s a deliberate decision.

@lorentzenchr not at all 😃 Feel free to propose an alternative

---

@aadya940 thank you for the updates!

Please keep in mind our CI when making updates. It needs manual approval for GitHub actions but there is still CircleCI and CirrusCI running every time you make an update.

You can add in your commit message one or both skip instructions: `[skip cirrus]` `[skip circle]`.

This is very important for us as our ressources are limited and shared with everyone.

One last thing, please take a moment to name your commits properly. You can have a look at our documentation here for some guidelines: https://scipy.github.io/devdocs/dev/contributor/development_workflow.html#writing-the-commit-message

No need to rewrite the history, in the end we use squash-merge. Also a force push breaks GitHub's UI so we ask people to not do that without asking us first.

**Comment by lorentzenchr** [comment]
Tensorflow also calls it softplus. That name seems to stem from [Dugas et al 2001](https://papers.nips.cc/paper_files/paper/2000/hash/44968aece94f667e4095002d140b5896-Abstract.html).
In that case, better keep the name `softplus`.

File: scipy/special/_basic.py
Comment (MatteoRaso) [inline scipy/special/_basic.py:3170]: I think this line would be confusing to users. It might be easier to understand if the possible keyword arguments were explicitly written, without reference to `np.logaddexp`.

**Comment by aadya940** [comment]
@person142 @rgommers

File: scipy/special/_basic.py
Comment (rgommers) [inline scipy/special/_basic.py:3189]: The return type is `ndarray` (or a scalar, but we document it as `ndarray` - you can copy how the `np.logaddexp` return value is documented).

File: scipy/special/_basic.py
Comment (rgommers) [inline scipy/special/_basic.py:3173]: minor: no blank line needed between parameters

File: scipy/special/tests/test_basic.py
Comment (rgommers) [inline scipy/special/tests/test_basic.py:3630]: Please use the `assert_allclose` function here.

File: scipy/special/_basic.py
Comment (rgommers) [inline scipy/special/_basic.py:3194]: can you please use `from scipy import special` here, as the idiomatic import?

File: scipy/special/_basic.py
Comment (rgommers) [inline scipy/special/_basic.py:3175]: `axis` and `axes` aren't keywords, and this won't render properly. I suggest copying exactly how `np.logaddexp` documents this, so it looks and renders like a ufunc.

File: scipy/special/tests/test_basic.py
Comment (rgommers) [inline scipy/special/tests/test_basic.py:3631]: Test coverage: can you please add a basic test using a couple of the common keywords? E.g.:
```python
x = np.arange(5) - 2
out = np.ones(5)
special.softplus(x, where=x>0, out=out)
assert_allclose(out, np.array([1., 1., 1., 1.31326169, 2.12692801]))
```

**Review by rgommers — DISMISSED** [review]
Thanks @aadya940. It seems a simple enough function to add, there was support on the linked issue and mailing list, so this seems close to ready. A few comments on docs and tests.

**Comment by aadya940** [comment]
Can anyone help me with what is breaking the build docs?

**Comment by tupui** [comment]
> Can anyone help me with what is breaking the build docs?

I would suggest to try reproduce the issue locally. From the log I am not sure what is happening. We have issues from time to time when the branch is far from `main`, but it does not seem to be the case here.

```
python dev.py doc
```

**Comment by aadya940** [comment]
> > Can anyone help me with what is breaking the build docs?
> 
> I would suggest to try reproduce the issue locally. From the log I am not sure what is happening. We have issues from time to time when the branch is far from `main`, but it does not seem to be the case here.
> 
> ```
> python dev.py doc
> ```

Couldn't find a solution, looks like its something from jsonschema . What might I be missing?

**Comment by tupui** [comment]
You are not missing anything :) there is actually an issue right now. Someone opened an issue about it so don't worry about it. Hopefully the issue will get resolved this week. (Locally you should be able to build with the previous version of jsonschema if you want).

**Comment by aadya940** [comment]
@rgommers

**Comment by aadya940** [comment]
What's the status of this PR? I guess the work is done.

**Comment by melissawm** [comment]
Hi @aadya940 - while we wait for a review here could you fix the conflicts that are now showing up? Apologies for the delay - most of our reviewers are volunteers and so it may take some time for them to be available.

**Comment by melissawm** [comment]
It looks like the linter picked up something:

```
scipy/special/tests/test_basic.py:44:27: F811 Redefinition of unused `mathieu_odd_coef` from line 42
scipy/special/tests/test_basic.py:44:45: F811 Redefinition of unused `mathieu_even_coef` from line 42
Found 2 errors.
```

@aadya940 would you take a look? Thanks!

**Comment by NeilGirdhar** [comment]
Would it be possible to make this work with the Array API?

File: scipy/special/tests/test_basic.py
Comment (mdhaber) [inline scipy/special/tests/test_basic.py:3728]: Good idea to test all those cases, but this doesn't quite do that:

image

Namely, `18 < x < 33.3` seems to be missing. Also the source of the reference values is not provided. I'll update accordingly.

File: scipy/special/tests/test_basic.py
Comment (mdhaber) [inline scipy/special/tests/test_basic.py:3743]: Let's generate the reference values programmatically. I'll update accordingly.

File: scipy/special/_basic.py
Comment (mdhaber) [inline scipy/special/_basic.py:3419]:
```suggestion
def softplus(x, **kwargs):
```

@rgommers [you asked](https://github.com/scipy/scipy/pull/18605/files/4dd4ddb55d2f2c32a931c9b4e0943104135fb3f9#diff-9a3516307ecc6b8a41e8631035448716fb36d7433c46a55a46b6f6ebebdd1483) for this to be documented exactly as `logaddexp` so it would look like a ufunc. 

I've implemented that change, but it complicates things a tiny bit if we want to make it array-API compatible (https://github.com/scipy/scipy/pull/18605#issuecomment-2046598660) because not all backends will accept `out` and `where`. Shall we just accept `**kwargs` rather than `out` and `where` specifically and remove the specific documentation of `out` and `where` below, just referring users to the ufunc docs for kwargs?

**Comment by mdhaber** [comment]
@aadya940 Are you still happy with this? I'll fix the CI failures, which just need a small tolerance bump.

Note that a name like `log1pexp` was [also suggested](https://github.com/scipy/scipy/pull/18605#issuecomment-1571943580) for this function; around the same time, gh-19021 suggested a `log1mexp` function. We could consider combining them with a `sign` argument (the sign of `exp(x)`) or a weight argument like `b` of `logsumexp`. However, perhaps it's OK to keep them separate, since the motivations/use cases are pretty different?

File: scipy/special/tests/test_basic.py
Comment (mdhaber) [inline scipy/special/tests/test_basic.py:3712]:
```suggestion
        assert_allclose(res, ref, rtol=2e-15)
```

**Comment by aadya940** [comment]
> @aadya940 Are you still happy with this? I'll fix the CI failures, which just need a small tolerance bump.
> 
> Note that a name like `log1pexp` was [also suggested](https://github.com/scipy/scipy/pull/18605#issuecomment-1571943580) for this function; around the same time, [gh-19021](https://github.com/scipy/scipy/issues/19021) suggested a `log1mexp` function. We could consider combining them with a `sign` argument (the sign of `exp(x)`) or a weight argument like `b` of `logsumexp`. However, perhaps it's OK to keep them separate, since the motivations/use cases are pretty different?

Keeping a sign argument seems a good idea IMO :) . Please let me know if I can help.

**Comment by mdhaber** [comment]
> Keeping a sign argument seems a good idea IMO

But then that raises the issue of naming the function again. Is it common for the "softplus" function to have a sign argument?

It also complicates the implementation - we can't just delegate to `np.logaddexp` (with real arguments, anyway).

**Comment by mdhaber** [comment]
Re: implementation where `sign` is negative, see gh-19201. You get better precision if you branch depending on the value.

And since this is an elementwise function, `sign` should be allowed to be an array, too. So you can't just branch with an `if` statement in Python; you would either need to use `_lazywhere` (slow) or compile this as a ufunc. 

Re: 
> ValueError("log(1 - e^x) is undefined for x >= 0")

Rather than emitting an error, a ufunc would probably return NaN in this case. Other options would include returning a value and sign of the result as separate outputs, or allowing the output to be complex. (The log of `-x` is `log(x) + np.pi*1j`). Personally, I would prefer to just return NaN. If the user wants full flexibility for complex input/output, they can use `logsumexp`.

So those are reasons why I wrote that it complicates the implementation. It's all simple, but it is much more complicated than what is here.

**Comment by aadya940** [comment]
@mdhaber It looks like implementing `log1pexp` and `log1mexp` separately would make things much simpler considering the implementation details and dis-similarity in their use cases.  :))

File: scipy/special/_basic.py
Comment (steppi) [reply scipy/special/_basic.py:3419]: Personally, I'm fine with not putting `out` and `where` in the signature explicitly, and accepting only `**kwargs`, at least for now. Maybe there's something more that can be done down the line about `out` and `where` in the array API, but if so, I'm fine kicking that can down the road.

**Review by steppi — APPROVED** [review]
Everything looks good to me here, modulo the concern about `out` and `where` in other array backends. I don't think documenting this args matters enough now to force us to make a firm decision here.

The name pair `softplus` and `log1mexp` seems fine to me. If there was a nice, widely used name for `log1mexp`, it would make sense to use that too. I agree with @mdhaber, that having the name pair `log1pexp` and `log1mexp` would make the difference in typical use unclear, and could raise the question of why they are not combined.

File: scipy/special/_basic.py
Comment (mdhaber) [inline scipy/special/_basic.py:3461]:
```suggestion
    return np.logaddexp(0, x, **kwargs)
```

File: scipy/special/_basic.py
Comment (mdhaber) [inline scipy/special/_basic.py:3441]:
```suggestion
```

**Comment by mdhaber** [comment]
Thanks @aadya940 et al!
@NeilGirdhar saw your comments about adding the inverse and making this array API compatible. PRs welcome!

**Comment by h-vetinari** [comment]
Was just updating the [release notes](https://github.com/scipy/scipy/wiki/*-Release-Note-Entries-for-SciPy-1.15.0-(asterisk-makes-this-appear-at-the-top)) for the factorial changes, and noticed that this doesn't have an entry yet.

**Pull request merged**
